### ISSUE TITLE
Watcher API re-connect issues

### ISSUE THREAD

**Issue by dpastorc** [etcd-cpp-apiv3/etcd-cpp-apiv3#117]
Hi,

First, my apologies for this long comment.

I have read previous issues #73 and #76 but I think that there are still issues related to watcher re-connect.

### Documentation Issues
The first issue is related to documentation at Readme.md file. I think that the current code isn't right:
```
void wait_for_connection(Client &client) {
  // wait until the client connects to etcd server 
  // `head` API is only available in version later than 0.2.1
  while (!client.head().get().is_ok()) {
    sleep(1);
  }
}

void initialize_watcher(const std::string &endpoints, 
  const std::string &prefix,
  std::function<void(Response)> callback, 
  std::shared_ptr<etcd::Watcher> &watcher) {
    Client client(endpoints);
    wait_for_connection(client); 
    watcher->reset(new etcd::Watcher(client, prefix, callback));
    watcher->Wait([endpoints, prefix, callback,
                   watcher_ref /* keep the shared_ptr alive */, &watcher](bool cancelled) {
        if (cancelled) {
            return;
        }
        initialize_watcher(endpoints, prefix, callback, watcher);
    });
}

std::string endpoints = "http://127.0.0.1:2379";  
std::function<void(Response)> callback = printResponse;
const std::string prefix = "/test/key"; 

// the watcher initialized in this way will auto re-connect to etcd
std::unique_ptr<etcd::Watcher> watcher;
initialize_watcher(endpoints, prefix, callback, watcher);
```
I have found the following issues:

1. Type error calling watch reset, it should be `watcher.reset(new etcd::Watcher(...)` instead `watcher->reset(new etcd::Watcher(...)`
2. Typo error in lambda parameters?
```
watcher->Wait([endpoints, prefix, callback,
		watcher_ref /* keep the shared_ptr alive */, &watcher](bool cancelled) { };
```
Compiler error on `watcher_ref`. What do you really want here?
Do you want to capture `watcher` variable by reference in the lambda function?
```
watcher->Wait([endpoints, prefix, callback,
		/*watcher_ref keep the shared_ptr alive */ &watcher](bool cancelled) {
	if (cancelled) {
		return;
	}
	initialize_watcher(endpoints, prefix, callback, watcher);
});
```
or, Do you want capture `watcher` variable by value in the lambda function?
```
// Notice that in this case lambda requires `mutable` qualifier.
watcher->Wait([endpoints, prefix, callback,
		watcher_ref  = /*keep the shared_ptr alive */ watcher](bool cancelled) mutable {
	if (cancelled) {
		return;
	}
	initialize_watcher(endpoints, prefix, callback, watcher_ref);
});
```

These two options will make the code to compile successfully, but they have different consequences. I will explain it later.

3. To be able compile it, you have to use a `shared_ptr` instead `unique_ptr` when `initialize_watcher()` is called for first time, so code should be:
```
// the watcher initialized in this way will auto re-connect to etcd
std::shared_ptr<etcd::Watcher> watcher;
initialize_watcher(endpoints, prefix, callback, watcher);
```

I think that some of these issues has been introduced in commits 2338a83 (#76) and 891dc7b

### Behaviour Issues
Related to re-connect behaviour (#73), I think that is still not fixed. In my case, I need to use a re-connect watcher but I would like to cancel the watcher when unlock operation happens. Depending the type of capture over `watch` variable used in lambda function, I have different behaviours but neither fits the objective.
#### Using capture by reference
Running the following test code:
```
void initialize_watcher(const std::string& endpoints,
	const std::string& prefix,
	std::function<void(etcd::Response)> callback,
	std::shared_ptr<etcd::Watcher>& watcher) {
	etcd::Client client(endpoints);
	original::wait_for_connection(client);
	watcher.reset(new etcd::Watcher(client, prefix, callback));
	watcher->Wait([endpoints, prefix, callback,
		/*watcher_ref  keep the shared_ptr alive ,*/ &watcher](bool cancelled) {
			if (cancelled) {
				return;
			}
			original::initialize_watcher(endpoints, prefix, callback, watcher);
		});
}

std::string my_endpoints = "http://127.0.0.1:2379";
std::function<void(etcd::Response)> callback = ot_etcd::print_event_response;
const std::string my_prefix = "/test/key";

// the watcher initialized in this way will auto re-connect to etcd
std::shared_ptr<etcd::Watcher> watcher;
original::initialize_watcher(my_endpoints, my_prefix, callback, watcher);

std::this_thread::sleep_for(std::chrono::seconds(10));
std::cout << "sleep one completed" << std::endl;

watcher->Cancel();
std::this_thread::sleep_for(std::chrono::seconds(10));
std::cout << "sleep two completed" << std::endl;
```
In this case there are two results:
1. The connection isn't lost during the tests. In this case the `watcher->Cancel()` is completed, so the watcher won't report anything after cancelling and all will work as expected.
2. But, if the connection is lost after `initialize_watcher()` call and before teh `Cancel()` call (I'm forcing it closing socket externally) then an exception happens. More precisely when `watcher.reset()` is executed  then `etcd::Watcher` destructor of previous object is called internally that leads throwing `_Throw_Cpp_error(_RESOURCE_DEADLOCK_WOULD_OCCUR);` due the `thread::join()` is executed from the same thread that `wait_callback` is called in `Watcher::doWatch()`.

#### Using capture by value
Running the following test code:
```
void initialize_watcher(const std::string& endpoints,
	const std::string& prefix,
	std::function<void(etcd::Response)> callback,
	std::shared_ptr<etcd::Watcher>& watcher) {
	etcd::Client client(endpoints);
	original::wait_for_connection(client);
	watcher.reset(new etcd::Watcher(client, prefix, callback));

	// Notice that lambda requires `mutable`qualifier.
	watcher->Wait([endpoints, prefix, callback,
		/*watcher_ref  keep the shared_ptr alive ,*/ watcher](bool cancelled) mutable {
			if (cancelled) {
				return;
			}
			original::initialize_watcher(endpoints, prefix, callback, watcher);
		});
}

std::string my_endpoints = "http://127.0.0.1:2379";
std::function<void(etcd::Response)> callback = ot_etcd::print_event_response;
const std::string my_prefix = "/test/key";

// the watcher initialized in this way will auto re-connect to etcd
std::shared_ptr<etcd::Watcher> watcher;
original::initialize_watcher(my_endpoints, my_prefix, callback, watcher);

std::this_thread::sleep_for(std::chrono::seconds(10));
std::cout << "sleep one completed" << std::endl;

watcher->Cancel();
std::this_thread::sleep_for(std::chrono::seconds(10));
std::cout << "sleep two completed" << std::endl;
```
In this case, when we capture `std::shared_ptr` by value then a new `std::shared_ptr` is created when lambda captures by value and increments the reference count.
Again there are two results:
1. The connection isn't lost during the tests. In this case the `watcher->Cancel()` is completed, so the watcher won't report anything after cancelling and all will work as expected.
2. But, if the connection is lost after `initialize_watcher()` call and before the `Cancel()` call ((I'm forcing it closing socket externally), then code doesn't crash by exception (it doesn't because the previous `shared_ptr` isn't release its object because reference count increased when parameter was capture by value in lambda). But `watcher->Cancel()` doesn't have effect due we are calling `Cancel` on the old object, then we are continue receiving notifications from a watcher (the last `Watcher` object created).
So with this approach, we will have a dangled Watcher object each time that a connection lost happens and new `Watcher` is be created in `initialize_watcher()`.

Probably the right way is to capture by reference in lambda, but then we have issues destroying `Watch` on cancel. On the other hand, avoiding `thread:join()` issues capturing by value in lambda makes that we will have dangled objects and cancellation doesn't work either, although app doesn't crash.
How to fix that? Any ideas?

Thanks.

**Comment by dpastorc** [comment]
I have a fix, see pull-request (#118) for details.

**Comment by sighingnow** [comment]
See also https://github.com/etcd-cpp-apiv3/etcd-cpp-apiv3/pull/118#issuecomment-1092775690
